src/main/java/com/github/fppt/jedismock/datastructures/streams/StreamId.java

```diff
@@ -0,0 +1,145 @@
+package com.github.fppt.jedismock.datastructures.streams;
+
+import com.github.fppt.jedismock.datastructures.Slice;
+import com.github.fppt.jedismock.exception.WrongStreamKeyException;
+
+import static com.github.fppt.jedismock.datastructures.streams.StreamErrors.RANGES_END_ID_ERROR;
+import static com.github.fppt.jedismock.datastructures.streams.StreamErrors.RANGES_START_ID_ERROR;
+import static java.lang.Long.compareUnsigned;
+import static java.lang.Long.parseUnsignedLong;
+import static java.lang.Long.toUnsignedString;
+import static com.github.fppt.jedismock.datastructures.streams.StreamErrors.INVALID_ID_ERROR;
+import static com.github.fppt.jedismock.datastructures.streams.StreamErrors.ZERO_ERROR;
+
+public class StreamId implements Comparable<StreamId> {
+    private final long firstPart;
+    private final long secondPart;
+
+    public long getFirstPart() {
+        return firstPart;
+    }
+
+    public long getSecondPart() {
+        return secondPart;
+    }
+
+    /* 0-0 ID for PRIVATE API usage */
+    StreamId() {
+        this(0, 0);
+    }
+
+    /* ID from long numbers for PRIVATE API usage */
+    public StreamId(long firstPart, long secondPart) {
+        this.firstPart = firstPart;
+        this.secondPart = secondPart;
+    }
+
+    public StreamId(String key) throws WrongStreamKeyException {
+        String[] parsed = key.split("-");
+
+        switch (parsed.length) {
+            case 2:
+                try {
+                    secondPart = parseUnsignedLong(parsed[1]);
+                } catch (NumberFormatException e) {
+                    throw new WrongStreamKeyException(INVALID_ID_ERROR);
+                }
+
+                try {
+                    firstPart = parseUnsignedLong(parsed[0]);
+                } catch (NumberFormatException e) {
+                    throw new WrongStreamKeyException(INVALID_ID_ERROR);
+                }
+
+                break;
+            case 1:
+                secondPart = 0;
+
+                try {
+                    firstPart = parseUnsignedLong(parsed[0]);
+                } catch (NumberFormatException e) {
+                    throw new WrongStreamKeyException(INVALID_ID_ERROR);
+                }
+
+                break;
+            default:
+                throw new WrongStreamKeyException(INVALID_ID_ERROR);
+        }
+    }
+
+    public StreamId(Slice slice) throws WrongStreamKeyException {
+        this(slice.toString());
+    }
+
+    public StreamId compareToZero() throws WrongStreamKeyException {
+        if (secondPart == 0 && firstPart == 0) {
```

Comment:
AFAICS the only use of this in source code is catching and rethrowing the result

Please rename it and rewrite it to `boolean isZero()` so that exception is not used to control flow

Reply:
The idea of such implementation is that when in XADD we are trying to create StreamID:
```java
try {
     nodeId = compareWithTopKey(new StreamId(stream.replaceAsterisk(id)).compareToZero());
} catch (WrongStreamKeyException e) {
     return Response.error(e.getMessage());
}
```

constructor might throw an exception as well as `compareWithTopKey` method. So I've writen this way to simplify code and follow the same style

Reply:
Your `compareWithTopKey` and `compareToZero` methods (note the arbitrary usage of 'with' and 'to') actually return their argument or throw an exception. So, they are not transforming their argument but just validating it. For this scenario, in Java practice, methods in the form of `void validateFoo(...) throws Bar` are used. This code can be rewritten in a clearer way without shifting the responsibility of validation to the `StreamId` class (see code below). And even this variant is far from perfect. If we can throw fewer exceptions in Java, it's better to do so. Each exception carries a call stack (leading to heap & garbage collector overload and CPU time to collect the call stack). No compiler optimizations are made for exceptional branches of code -- intentionally.

```java
//package-private, not public, for testing purposes
public void validate(StreamId key) throws WrongStreamKeyException {
        if (key.isZero()) { //no RMStream is needed at this point, so let's check it first
            throw new WrongStreamKeyException(ZERO_ERROR);
        }
        StreamId lastId = getStreamFromBaseOrCreateEmpty(params().get(0)).getLastId();
        if (key.compareTo(lastId) <= 0) {
            throw new WrongStreamKeyException(TOP_ERROR);
        }
    }


protected Slice response() {
   //...blah- blah...
        StreamId nodeId;
        try {
            nodeId = new StreamId(stream.replaceAsterisk(id));
            validate(nodeId);
        } catch (WrongStreamKeyException e) {
            return Response.error(e.getMessage());
        }
  //...
}
```